docs/proposals/20210203-unmanaged-infrastructure.md

```diff
@@ -0,0 +1,235 @@
+- Support anything other than cluster infrastructure (e.g. machines).
+
+## Proposal
+
+A new annotation `cluster.x-k8s.io/managed-by: "<name-of-system>"` is going to be defined in Cluster API core repository, which helps define and identify resources managed by external controllers. The value of the annotation will not be checked by Cluster API and is considered free form text.
+
+Infrastructure providers SHOULD respect the annotation and its contract.
+
+When this annotation is present on an InfraCluster resource, the InfraCluster controller is expected to ignore the resource and not perform any reconciliation.
+Importantly, it will not modify the resource or its status in any way.
+A predicate will be provided in the Cluster API repository to aid provider implementations in filtering resources that are externally managed.
+
+Additionally, the external management system must provide all required fields within the spec of the InfraCluster and must adhere to the CAPI provider contract and set the InfraCluster status to be ready when it is appropriate to do so.
+
+While an "externally managed" InfraCluster won't reconcile or manage the lifecycle of the cluster infrastructure, CAPI will still be able to create compute nodes within it.
+
+The machine controller must be able to operate without hard dependencies regardless of the cluster infrastructure being managed or externally managed.
+![](https://i.imgur.com/nA61XJt.png)
+
+### User Stories
+
+#### Story 1 - Alternate control plane provisioning with user managed infrastructure
+As a cluster provider I want to use CAPI in my service offering to orchestrate Kubernetes bootstrapping while letting workload cluster operators own their infrastructure lifecycle.
+
+#### Story 2 - Restricted access to cloud provider APIs
+As a cluster operator I want to use CAPI to orchestrate kubernetes bootstrapping while restricting the privileges I need to grant for my cloud provider because of organisational cloud security constraints.
+
+#### Story 3 - Consuming existing cloud infrastructure
+As a cluster operator I want to use CAPI to orchestate Kubernetes bootstrapping while reusing infrastructure that has already been created in the organisation either by me or another team.
+
+### Implementation Details/Notes/Constraints
+
+**Managed**
+
+- It will be default and will preserve existing behaviour. An InfraCluster CR without the `cluster.x-k8s.io/managed-by: "<name-of-system>"` annotation.
+
+
+**Externally Managed**
+
+An InfraCluster CR with the `cluster.x-k8s.io/managed-by: "<name-of-system>"` annotation.
+
+The provider InfraCluster controller must:
+- Skip any reconciliation of the resource.
+
+- Not update the resource or its status in any way
+
+The external management system must:
+
+- Populate all required fields within the InfraCluster spec to allow other CAPI components to continue as normal.
+
+- Adhere to all Cluster API contracts for infrastructure providers.
```

Comment:
Is there anything special we have to consider regarding cluster deletion? I think it should by pretty similar to cluster creation, so I assume this will work out.

Reply:
Cluster resource deletion will enforce deletion ordering of resources as usual and the infraProvider CR should just no-op, same as creation.